### ISSUE TITLE
Redact password in logs

### ISSUE THREAD

**Issue by mleonhard** [servo/rust-url#714]
Software often logs urls using std::format::Display.  Example from [ureq HTTP client library](https://github.com/algesten/ureq/blob/2f1f9c45aab32cdcf74668473c4ec33e3d16d215/src/unit.rs#L242):

```rust
info!("sending request {} {}", method, url);
```

When the Url contains a password, the password ends up in the logs:

```
INFO sending request GET https://ACa6e1ce:5eadb27757bd50f8f01d4bcac866e@lookups.twilio.com/v1/PhoneNumbers/10001112222
```

This is a security risk.  Let's change Url's impl Display to redact the password:
```
INFO sending request GET https://ACa6e1ce:***@lookups.twilio.com/v1/PhoneNumbers/10001112222
```

**Comment by tmccombs** [comment]
That could potentially break code that assumes the `Display` implementation  produces the same string as `as_str`. Also note that that could also call from calling [`to_string`](https://doc.rust-lang.org/stable/std/string/trait.ToString.html#tymethod.to_string) since the `ToString` trait is implemented for anything that implements `Display`.

I think it would generally be good... but probably isn't backwards compatible.

**Comment by dpc** [comment]
Any recommendations for a library that would do that? All is needed is some kind of a newtype wrapping `Url` and changing `Display` & `Debug` impls.

**Comment by nyurik** [comment]
I agree this is a frequent issue for anyone trying to print/log a sensitive URL. Note that hiding username/password from a URL may not be enough for some cases. For example, a URL accessing an AWS S3 bucket may contain secrets in the query string (see [docs](https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html)):

```
https://s3.amazonaws.com/examplebucket/test.txt
?X-Amz-Algorithm=AWS4-HMAC-SHA256
&X-Amz-Credential=<your-access-key-id>/20130721/us-east-1/s3/aws4_request
&X-Amz-Date=20130721T201207Z
&X-Amz-Expires=86400
&X-Amz-SignedHeaders=host
&X-Amz-Signature=<signature-value>  
```

I wonder if URL sanitation is a complex-enough problem to warrant its own crate...

**Comment by valenting** [comment]
> I wonder if URL sanitation is a complex-enough problem to warrant its own crate...

I think so.
As noted, the solution can be a simple wrapper type, but as you state there are lots of ways secrets can be embedded in URLs.
I won't rule out adding this feature to the url crate, but considering this issue is not a concern for all users of the crate and that it can be resolved with the addition of a separate crate, I think that's a better option for the moment.

**Comment by abonander** [comment]
This is a perfect use-case for the [alternative representation flag](https://doc.rust-lang.org/stable/std/fmt/struct.Formatter.html#method.alternate): 

```rust
let url: Url = "https://foo:bar@quux:1234/path?key=asdf1234".parse().unwrap();

assert_eq!(url.to_string(), "https://foo:bar@quux:1234/path?key=asdf1234");
assert_eq!(format!("{url:#}", "https://foo:REDACTED@quux:1234/path?REDACTED");
```

**Comment by dpc** [comment]
IMO ideally, the default would be redacted, and opt-in would be used when this is undesireable.

BUT also ... this is probably too implicit and easy miss and break.

Probably a separate method that returns a custom formatting newtype wrapper that does not redact would make the decision of printing including all details more clear.

**Comment by Scripter17** [comment]
> IMO ideally, the default would be redacted, and opt-in would be used when this is undesireable.
> 
> BUT also ... this is probably too implicit and easy miss and break.
> 
> Probably a separate method that returns a custom formatting newtype wrapper that does not redact would make the decision of printing including all details more clear.

1. Having `Url::parse(url.as_str())` not roundtrip is bad (glares at `Url::set_path` letting URLs end in opaque paths that end in a literal space)
2. Having `url.to_string()` not be the same as `url.as_str().to_string()` is bad (obviously)
3. `ToString` is automatically implemented for all `T: Display` using `format!("{self}")` (even if you could decouple them, it'd be surprising)

A method like `display_redacted<T: RedactionPolicy = RedactPassword>(&self) -> ...` is probably the most comprehensive and least noisy option? It allows redacting non-password parts as needed while allowing the above `Display::alternate` suggestion.

The exact API is too close to the question of "is it even possible to make a good URL API" for me to want to think much about right now. The main issue is if it should take a `&T` or be purely generic based.

There is also the question of if that level of complexity is even a good fit given how bare bones the rest of the API is.

**Comment by abonander** [comment]
I don't think it's necessary to make this super generic or pluggable, if the user needs more flexibility this could be implemented out-of-crate.

I threw together a quick proof-of-concept for how simple this API can be: https://github.com/servo/rust-url/compare/main...abonander:rust-url:ab/redacted
